### ISSUE TITLE
Add PU-learning evaluation metrics

### ISSUE THREAD

**Issue by shaypal5** [pulearn/pulearn#93]
# Add PU-learning evaluation metrics (corrected confusion metrics, corrected AUC/PR-AUC, PU risk + diagnostics) to prevent misleading validation scores

## Background / Problem Statement
Issue #12 highlights a fundamental structural crisis in Positive-Unlabeled (PU) learning evaluation: traditional binary metrics (Precision, Recall, F1) are predicated on a balanced ground truth that includes confirmed negative examples. When validation is performed by holding out only labeled positives, a degenerate or "trivial" classifier that predicts the positive class for every instance ($f(x) = 1$) achieves a recall of 1.0. Without a mechanism to penalize false positives among the unlabeled pool, such models erroneously appear superior to discriminative models.

The pulearn library currently provides a minimal metrics module (src/pulearn/metrics.py) containing only recall and the lee_liu_score (a robust ranking proxy). To address the evaluation deficit observed in iterative learners like Adaboost, this suite must be expanded to provide unbiased estimators of true performance.

## PU Learning Evaluation: Assumptions and Constraints
The proposed metrics primarily operate under the Selected Completely At Random (SCAR) assumption. Under SCAR, the probability of a positive instance being labeled is a constant propensity score $c = P(s=1 | y=1)$, independent of features $x$. This allows for the conversion of observed probabilities into calibrated posteriors:

$$P(y=1|x) \approx P(s=1|x) / c$$

While metrics should assume SCAR as a baseline, documentation should note that Selection Bias (SAR/SNAR) requires more complex inverse propensity weighting.

### A) Core Utilities for Calibration
These functions enable the transition from observed labels ($s$) to estimated true labels ($y$).

- estimate_label_frequency_c(y_pu, s_proba)
       - Implements Elkan-Noto style estimation of the propensity score $c$.
       - Baseline: $c \approx \text{mean}(s\_proba[y\_pu == 1])$.
- calibrate_posterior_p_y1(s_proba, c_hat)
       - Clips and scales scores to the $$ range.
       - Logic: $p_{y1} = \text{clip}(s\_proba / c\_hat, 0, 1)$.

### B) Expected-Confusion Metrics (Unbiased F1 and Precision)
Rather than treating unlabeled data as negative, these metrics reconstruct the confusion matrix using the class prior $\pi$.

- Target Metrics: pu_precision_score, pu_recall_score, pu_f1_score, pu_specificity_score.
- Implementation Logic:
       - $TP \approx \sum p_{y1}$ (sum of probabilities for predicted positives).
       - $FP \approx \sum (1 - p_{y1})$ (expected negatives in the predicted positive set).
- Critical Goal: Ensure that "predict-all-positive" results in a specificity approaching 0.

### C) Ranking Metrics: Adjusted AUC and AUL
Ranking performance must be corrected to reflect the true $AUC_{pn}$ rather than the observed $AUC_{pu}$.

- pu_roc_auc_score(y_true, y_score, pi)
       - Implements the Sakai 2018 adjustment:
       - $$AUC_{pn} = \frac{AUC_{pu} - 0.5\pi}{1 - \pi}$$
- pu_average_precision_score
       - Incorporates the Area Under Lift (AUL), which is linearly related to AUC but more robust to class imbalance.
       - $$AUL = 0.5\alpha + (1 - \alpha)AUC$$

### D) Risk Minimization and Proper Scoring
For flexible models (e.g., Deep Learning), raw risk estimators are necessary for early stopping and model selection.

- pu_unbiased_risk (uPU)
       - Implements $\hat{R}_{pu}(g) = \pi \hat{R}_p^+(g) + \hat{R}_u^-(g) - \pi \hat{R}_p^-(g)$.
- pu_non_negative_risk (nnPU)
       - Enforces a non-negative constraint on the negative risk component to prevent overfitting.

### E) Diagnostics and Distribution Alignment
Advanced guardrails to detect if a model is failing to distinguish hidden positives.

- pu_distribution_diagnostics
       - Calculates the Kullback-Leibler (KL) Divergence between $P(f(x)|s=1)$ and $P(f(x)|s=0)$.
- homogeneity_metrics
       - Reports the Standard Deviation (STD) and Inter-Quartile Range (IQR) of probabilities for identified negatives to detect over-reliance on trivial features.

### F) Scikit-Learn Integration
Provide a factory function make_pu_scorer that allows these metrics to be used in GridSearchCV or RandomizedSearchCV.

- Signature: make_pu_scorer(metric_name, pi,...)
- Functionality:
       - Should wrap the PU logic using sklearn.metrics.make_scorer to ensure compatibility with scoring='pu_f1' style calls.

### Update src/pulearn/metrics.py
- Preserve existing recall and lee_liu_score functions.
- Add new functions grouped by:
       - Calibration
       - Confusion
       - Ranking
       - Risk
       - Diagnostics

### Create tests/test_pu_metrics_corrected.py
- SCAR Parity Test
       - Assert that PU-corrected metrics on simulated SCAR data are approximately equal to supervised metrics computed on ground-truth labels.
- Degenerate Case Test
       - Verify that a constant "positive" predictor yields specificity $\approx 0$.

### Add examples/PUMetricsEvaluationExample.py
- Demonstrate the failure of naive F1 vs. the success of the Li-Liu score and Unbiased F1 on a standard dataset (e.g., Breast Cancer or MNIST).

## Acceptance Criteria
- pulearn.metrics exposes the full suite while maintaining backward compatibility.
- All new metrics support scikit-learn model selection workflows.
- Clear documentation exists for assumption requirements (e.g., providing $\pi$).
- Test coverage remains at 100% for the metrics module.

# Background: Engineering Robust Evaluation Metrics for Positive-Unlabeled Learning Frameworks  
## Introduction

- The computational paradigm of Positive-Unlabeled (PU) learning represents a critical frontier in semi-supervised binary classification, addressing scenarios where explicit negative supervision is absent.
- In such environments, the training corpus consists of:
       - A small set of confirmed positive instances.
       - A larger pool of unlabeled data that inherently aggregates both hidden positives and true negatives.
- This structural asymmetry poses a significant challenge for traditional model evaluation methodologies.
- Standard metrics such as precision, accuracy, and F1-score are fundamentally predicated on the existence of a balanced ground truth, including a well-defined negative class.
- When these metrics are naively applied to PU data—typically by treating the unlabeled set as a proxy for the negative class—the resulting performance estimates are systematically biased and often misleading.
- This report provides:
       - An exhaustive examination of the evaluation deficit identified in pulearn issue #12.
       - A characterization of theoretical advancements in PU-specific metrics.
       - A comprehensive implementation strategy for integrating a robust metric suite into the package.
- By synthesizing foundational research with recent breakthroughs in unbiased risk estimation and metric learning, this analysis establishes the technical foundation required for automated development agents, such as GitHub CoPilot, to implement these solutions effectively.

## Theoretical Frameworks and the Ontology of the PU Evaluation Crisis

- The fundamental difficulty in PU learning evaluation stems from the inability to directly observe false positives and true negatives.
- In standard binary classification:
       - We operate on a joint distribution $p(x, y)$.
       - $y \in \{0, 1\}$.
- In the PU setting:
       - Introduce latent observation variable $s \in \{0, 1\}$.
       - $s = 1$ indicates labeled.
       - $s = 0$ indicates unlabeled.
- Under the standard "perfect precision" assumption:
       - $p(y=1 \mid s=1) = 1$.
- The unlabeled set satisfies:
       - $p(y=1 \mid s=0) > 0$.
- Evaluation historically relies on the SCAR assumption:
       - $p(s=1 \mid x, y=1) = p(s=1 \mid y=1) = c$.
       - $c$ is the propensity score.
- In real-world data:
       - Selection Bias or SAR may apply.
       - Propensity score $e(x)$ depends on features.
- Failure of standard metrics:
       - Is structural.
       - Allows trivial classifiers to appear optimal.

## Analysis of the pulearn Evaluation Deficit (Issue #12)

- Practitioners often:
       - Hold out labeled positives for validation.
- If classifier predicts:
       - $f(x) = 1$ for all samples.
- Then:
       - Recall = 1.0 on held-out positives.
- Without penalizing predictions on unlabeled data:
       - "All-positive" classifiers appear optimal.
- Root cause:
       - No penalization for false positives when unlabeled data is ignored.
- Resolution:
       - Replace traditional binary metrics with PU-derived metrics.

### Structural Comparison

- Feature: Negative Samples
       - Standard: Explicit, used for Specificity and FP.
       - PU: Absent; unlabeled contains noise.
- Feature: Positive Samples
       - Standard: Used for Recall and TP.
       - PU: Only biased subset available.
- Feature: Class Prior ($\pi$)
       - Standard: Determined from sample.
       - PU: Must be estimated or assumed.
- Feature: Metric Bias
       - Standard: Unbiased under IID.
       - PU: Systematically biased.

### 1. Li and Liu Criterion (F1 Proxy)

- Designed for model selection without negative labels.
- Formula:
       - $C = \frac{r^2}{Pr[f(X)=1]}$
- Where:
       - $r$ is recall on positives.
       - $Pr[f(X)=1]$ is fraction predicted positive.
- Properties:
       - Penalizes over-prediction.
       - Requires no class prior $\pi$.
- Example:
       - All-positive: $1.0$.
       - Specific classifier: $0.9^2 / 0.1 = 8.1$.
- Primary use:
       - Cross-validation and model ranking.

### 2. Unbiased and Non-Negative Risk Estimators (uPU, nnPU)

- True risk:
       - $R(g) = \pi R_p^+(g) + (1-\pi)R_n^-(g)$.
- Replace negative risk via:
       - $R_u^- - \pi R_p^-$.
- Empirical estimator:
       - $\hat{R}_{pu}(g) = \pi \hat{R}_p^+(g) + \hat{R}_u^-(g) - \pi \hat{R}_p^-(g)$.
- Issues:
       - Negative risk values possible.
- nnPU:
       - Enforces non-negativity constraint.
- Use:
       - Early stopping and validation in flexible models.

### 3. Ranking Metrics

- PU-AUC:
       - $AUC_{pn} = \frac{AUC_{pu} - 0.5\pi}{1 - \pi}$.
       - Valid surrogate under SCAR.
- Area Under Lift (AUL):
       - Measures ranking of positive vs population.
       - Lower estimation error than AUC.
- G-mean:
       - Geometric mean for balanced evaluation.

## Software Engineering Requirements for pulearn.metrics

- Create:
       - `src/pulearn/metrics.py`.
- Ensure:
       - Full scikit-learn API compatibility.
       - Use of make_scorer wrappers.
- Implement:
       - SCORERS dictionary.

### 1. li_liu_score

- Input:
       - y_true
       - y_pred
- Steps:
       - pos_indices = y_true == 1
       - recall = mean(y_pred[pos_indices] == 1)
       - positive_prob = mean(y_pred == 1)
       - return recall**2 / positive_prob if positive_prob > 0 else 0.0

### 2. unbiased_f1_score

- Input:
       - y_true
       - y_pred
       - pi
- Steps:
       - recall = mean(y_pred[y_true == 1] == 1)
       - precision = (pi * recall) / mean(y_pred == 1)
       - precision = clip(precision, 0, 1)
       - return harmonic mean if denominator > 0 else 0

### 3. pu_auc_score

- Input:
       - y_true
       - y_score
       - pi (optional)
- Steps:
       - raw_auc = roc_auc_score(y_true, y_score)
       - if pi is None:
              - return raw_auc
       - else:
              - return (raw_auc - 0.5 * pi) / (1 - pi)

### 4. Diagnostic Metrics

- Kullback-Leibler Divergence:
       - Compare distributions of predicted probabilities.
- STD / IQR of predicted negatives:
       - Detect homogeneity issues.
- Purpose:
       - Detect algorithmic bias.
       - Validate diversity of negative representation.

## Scorer Interface Architecture

- Metric: li_liu_score
       - Greater is Better: True
       - Needs Proba: False
       - Signature: (y_true, y_pred)
- Metric: unbiased_f1
       - Greater is Better: True
       - Needs Proba: False
       - Signature: (y_true, y_pred, pi)
- Metric: pu_auc
       - Greater is Better: True
       - Needs Proba: True
       - Signature: (y_true, y_score, pi)
- Metric: negative_risk
       - Greater is Better: False
       - Needs Proba: True
       - Signature: (y_true, y_score, pi)

## Labeling Mechanisms and Metric Bias

- One-Sample (OS):
       - Unlabeled generated first.
       - Distribution: $p(x)$.
- Two-Sample (TS):
       - Positive and unlabeled sampled independently.
- Implication:
       - Metrics dependent on relative P/U sizes may be biased.
       - Requires correction using $\pi$.

## Propensity Score Calibration

- PU classifiers learn:
       - $p(s=1 \mid x)$.
- True probability:
       - $p(y=1 \mid x) = \frac{p(s=1 \mid x)}{c}$.
- Consequence:
       - Threshold shifts to $0.5 \cdot c$.
- Need:
       - Calibration methods (e.g., PULSCAR, PULSNAR).
- Without calibration:
       - F1 and threshold-dependent metrics flawed.

### Biomedical Applications

- Riboswitch identification.
- Use of ensemble PU classifiers.
- External synthetic ground-truth validation.
- Principle:
       - PU metrics for selection.
       - External validation for final assessment.

### Fraud Detection

- Extreme imbalance.
- Standard AUC insufficient.
- Recommend:
       - Risk re-weighting.
       - AUL metric.

### Social Media Analysis

- Politicization detection.
- Spy technique for reliable negatives.
- Spy recall:
       - Diagnostic metric for calibration.

### Module Structure

- Location:
       - `src/pulearn/metrics.py`
- Imports:
       - numpy
       - sklearn.metrics
       - sklearn.utils

### Strategy for All-Positive Adaboost Issue

- Override default scoring:
       - Use li_liu_score.
- Ensure:
       - GridSearchCV penalizes trivial classifiers.

## Future Outlook

- Move toward:
       - Metric Learning (e.g., P3M framework).
- Instance-Dependent PU learning:
       - Weight evaluation by inverse propensity $1/e(x)$.
- Future requirement:
       - Metrics accepting propensity vectors.

## Detailed Synthesis of Recommendations

- Highest priority:
       - Implement Li and Liu criterion.
- Secondary:
       - Unbiased F1 and Adjusted AUC.
- Add:
       - Diagnostic report utilities (KLD, STD, IQR).
- Ensure:
       - Full scikit-learn compatibility.
       - SCORERS dictionary integration.
- Outcome:
       - Establish pulearn as definitive PU evaluation toolkit.

## Summary of Metric Implementation Parameters

- Proxy (Li-Liu)
       - Complexity: Low
       - Requirement: None
       - Foundation: Lee & Liu (2003)
- Unbiased (uPU)
       - Complexity: Medium
       - Requirement: Class Prior $\pi$
       - Foundation: du Plessis et al. (2015)
- Non-Negative (nnPU)
       - Complexity: High
       - Requirement: Class Prior $\pi$
       - Foundation: Kiryo et al. (2017)
- Ranking (PU-AUC)
       - Complexity: Low
       - Requirement: Optional propensity score
       - Foundation: Sakai et al. (2018)
- Lift (AUL)
       - Complexity: Medium
       - Requirement: None
       - Foundation: Vuk & Curk (2006)

- The proposed suite addresses issue #12 while providing the theoretical depth required for advanced PU learning research and application in genomics, fraud detection, and beyond.
